File: advcake/lgcity/get_product.py
import random
from get_transliterate import get_transliterate
from selenium.webdriver.common.by import By
from selenium.webdriver.support.wait import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.options import Options
import undetected_chromedriver


def get_product(link, gender, idx):
    global category
    try:
        option = Options()
        option.headless = True
        option.add_argument("--headless=new")
        driver = undetected_chromedriver.Chrome(options=option)
        driver.get(link)
        try:
            if gender != 'women': gender = 'Мужской'
            else: gender = 'Женский'
            name = WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.ID, 'breadcrumbs-last-item'))).get_attribute('textContent')
            price = int(driver.find_element(By.CLASS_NAME, 'card__info-price-text--new').get_attribute('content'))
            oldprice = int(driver.find_element(By.CLASS_NAME, 'card__info-price-text--old').get_attribute('textContent')
                           .replace('\n', '').replace('\t', '').replace('\xa0', ' ').replace(' ', '').replace('₽', ''))
            sale = int(driver.find_element(By.CLASS_NAME, 'card__info-price-text--percent').get_attribute('textContent').replace('%', ''))
            brand = driver.find_element(By.CLASS_NAME, 'card__info-link').text.upper()
            category = driver.find_elements(By.CLASS_NAME, 'breadcrumbs__item')[2].get_attribute('textContent').strip()
            subcategory = driver.find_elements(By.CLASS_NAME, 'breadcrumbs__item')[4].get_attribute('textContent').strip()
            color = driver.find_element(By.CLASS_NAME, 'card__color-value').get_attribute('textContent').lower().replace('ё', 'е').replace('в наличии: онлайн, офлайн', '').replace('в наличии: онлайн', '')
            if color == '': color = False
            brand_country = False
            country = False
            season = False
            style = False
            structure = False
            keys = driver.find_elements(By.CLASS_NAME, 'card__info-list-item-key')
            values = driver.find_elements(By.CLASS_NAME, 'card__info-list-item-val')
            info = {}
            for idx, key in enumerate(keys):
                key_txt = key.get_attribute('textContent').replace(':', '')
                val_txt = values[idx].get_attribute('textContent')
                if key_txt == 'Состав': structure = val_txt
                if key_txt == 'Страна производства': country = val_txt
                if key_txt == 'Страна бренда': brand_country = val_txt
                if key_txt == 'Сезон': season = val_txt
                if key_txt == 'Стиль': style = val_txt
                info[key_txt] = val_txt
            sizes = []
            for size in driver.find_elements(By.CLASS_NAME, 'select__drop-list-text--right'):
                sizes.append(size.get_attribute('textContent'))
            images = []
            for img in driver.find_elements(By.CLASS_NAME, 'card__slider-img'):
                images.append(img.get_attribute('src'))
            print(idx, 'PRODUCT - DONE!!!')
        except Exception as ex:
            print(ex)
            return
        finally: driver.quit()
        return {
            'id': round(random.uniform(1000000000, 9999999999)),
            'age': 'Взрослый',
            'benefit': oldprice - price,
            'brand': brand,
            'brandCountry': brand_country,
            'brandCountry_t': get_transliterate(brand_country) if brand_country else False,
            'category': category,
            'category_t': get_transliterate(category),
            'color': color,
            'color_t': get_transliterate(color) if color else False,
            'country': country,
            'country_t': get_transliterate(country) if country else False,
            'delivery': ['ru'],
            'deliveryPrice': False,
            'gender': gender,
            'name': name,
            'info': info,
            'installment': False,
            'images': images,
            'like': 0,
            'link': link,
            'oldprice': oldprice,
            'pp': 'advcake',
            'price': price,
            'sale': sale,
            "season": season,
            'season_t': get_transliterate(season) if season else False,
            'shop': 'lgcity',
            'sizes': sizes,
            "style": style,
            'style_t': get_transliterate(style) if style else False,
            'structure': structure,
            'subcategory': subcategory,
            'subcategory_t': get_transliterate(subcategory)
        }
    except:
        print(f'{link} НЕ собран!')
        return


# print(get_product(
#     'https://lgcity.ru/product/bryuki_3887.html',
#     'Женский', 1))

File: admitad/brd_proto/getProduct.js
const axios = require('axios')
const { JSDOM } = require('jsdom')
const getOptions = require('./getOptions')
const getTransliterate = require('./getTransliterate')
const iconv = require('iconv-lite');


module.exports = async function getProduct (productLink, gender) {
  try {
    // получаю данные по товару
    const productOptions = getOptions(null, productLink)
    const productRes = await axios.post( ...productOptions )
    const ruHtmlProduct = iconv.decode(productRes.data, "win1251")
    const dom = new JSDOM(ruHtmlProduct).window.document

    const breadCrumbs = Array.from(dom.querySelectorAll('span[itemprop="name"]'))
      .map(crumb => crumb.textContent.toLowerCase())

    let category;
    breadCrumbs[2].indexOf('одежда') !== -1
      ? category = 'Одежда'
      : breadCrumbs[2].indexOf('обувь') !== -1
        ? category = 'Обувь'
        : category = 'Аксессуары'

    let subcategory = breadCrumbs[3].toLowerCase().replaceAll('мужской', '').replaceAll('мужская', '').replaceAll('мужские', '')
      .replaceAll('женский', '').replaceAll('женская', '').replaceAll('женские', '').trim()
    subcategory= subcategory[0].toUpperCase() + subcategory.slice(1)
    
    const brand = dom.querySelector('meta[itemprop=brand]').getAttribute('content').toUpperCase()

    const attributes = dom.querySelector('dl.attributes')
    colors = attributes.querySelectorAll('dd')

    const imageLinks = Array.from(dom.querySelector('section.desktop_images').querySelectorAll('a'))
      .map(link => link.href)

    const fullNameArr = dom.querySelector('header h1[itemprop="name"]').textContent.split(' ')
    let name = []
    function checkArr(fullName) {
        for (let word of fullName) {
          if (/^[0-9А-ЯЁ]+$/i.test(word)) name.push(word);
        }
      }
    checkArr(fullNameArr)
    name.push(brand)

    let infoDetails = dom.querySelector('div.details').querySelector('meta[itemprop="description"]').getAttribute("content").split('\n')

    let sale = parseInt(dom.querySelector('div.product-card__label.product-card__label--sale').textContent)
    if (sale < 0) {
        sale = sale * -1;
      }

    let sizes = [];
    let labels = Array.from(dom.querySelector('div.option.text').querySelectorAll('.options-item.js-ab-prinfo-add-popup-size-hover'))
    for (let label of labels) {
        if (label.getAttribute('class') == 'options-item js-ab-prinfo-add-popup-size-hover ') sizes.push(label);
      }
    sizes = Array.from(sizes).map(size => size.querySelector('span').textContent);

    const oldprice = parseInt(dom.querySelector('div.prices').querySelector('span.price_obsolete').textContent)
    const price = parseInt(dom.querySelector('div.prices').querySelector('span[itemprop="price"]').textContent)

    let structure = dom.querySelector('div.composition-info')
    // if (structure) structure = structure.textContent.split('/')
    if (structure) {
        structure = structure.textContent.split('/');
        structure[0] = structure[0].split(':')[1];
        for (let num in structure) {
            structure[num] = structure[num].trim();
          };
      } else {
        structure = false;
      }
    // structure[0] = structure[0].split(':')[1]
    // for (let num in structure) {
    //     structure[num] = structure[num].trim();
    //   }

    const product = {
      id: Math.floor(Math.random() * 9e9) + 1e9,
      age: 'Взрослый',
      benefit: oldprice - price,
      brand,
      brandCountry: false,
      brandCountry_t: false,
      category,
    //   category_t: getTransliterate(category),
      color: colors[colors.length - 1].textContent,
      color_t: getTransliterate(colors[colors.length - 1].textContent),
      country: false,
      country_t: false,
      create: new Date,
      delivery: ['ru'],
      deliveryPrice: false,
      description: false,
      gender: gender === 'men'
        ? 'Мужской'
        : 'Женский',
      installment: true,
      images: imageLinks,
      like: 0,
      link: productLink,
      name: name.join(' '),
      shop: 'brd',
      info: infoDetails,
      oldprice: oldprice,
      pp: 'admitad',
      price: price,
      sale: sale,
      season: false,
      season_t: false,
      sizes: sizes[0] == ''
        ? [ 'one size' ]
        : sizes,
      style: false,
      style_t: false,
      structure: structure,
      subcategory,
      subcategory_t: getTransliterate(subcategory),
    }
    return product
  } catch (e) {
    console.log('Товар', productLink, 'не собран', e)
    return null
  }
}
